File: examples/governance_static_go/domain/audit_log.go
package domain

import (
	"time"
	"github.com/google/uuid"
)

type AuditLog struct {
	ID string `json:"id"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
	EntityType string `json:"entity_type"`
	EntityId string `json:"entity_id"`
	Action string `json:"action"`
	ActorId string `json:"actor_id"`
	Details string `json:"details"`
	Timestamp time.Time `json:"timestamp"`
}

func NewAuditLog(entityType string, entityId string, action string, actorId string, details string, timestamp time.Time) *AuditLog {
	a := &AuditLog{
		ID:        uuid.New().String(),
		EntityType: entityType,
		EntityId: entityId,
		Action: action,
		ActorId: actorId,
		Details: details,
		Timestamp: timestamp,
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}
	return a
}

func (a *AuditLog) Validate() error {
	if a.EntityType == "" {
		return &ValidationError{Field: "entity_type", Message: "entity_type can't be blank"}
	}
	if a.Action == "" {
		return &ValidationError{Field: "action", Message: "action can't be blank"}
	}
	return nil
}

File: examples/governance/identity_domain/lib/identity_domain/audit_log/audit_log.rb
require 'hecks/mixins/model'

module IdentityDomain
  class AuditLog
    include Hecks::Model

    attribute :entity_type
    attribute :entity_id
    attribute :action
    attribute :actor_id
    attribute :details
    attribute :timestamp

    private

    def validate!
      raise ValidationError.new("entity_type can't be blank", field: :entity_type, rule: :presence) if entity_type.nil? || (entity_type.respond_to?(:empty?) && entity_type.empty?)
      raise ValidationError.new("action can't be blank", field: :action, rule: :presence) if action.nil? || (action.respond_to?(:empty?) && action.empty?)
    end
  end
end
